### PR TITLE
Canvas: Add element snapping and alignment

### PR DESCRIPTION
**Pull request by nmarrs**
There are a few potentially low hanging fruit that we should take advantage of with improving canvas.

One of these is the [snappable feature](https://daybrush.com/moveable/storybook/index.html?path=/story/snap-bound--snap-guidelines) built into the moveable library. This PR pushes canvas further into a more matured product experience / helps smooth over its rougher edges.

https://github.com/grafana/grafana/assets/22381771/6e8aa5ad-7dcc-432e-ac2e-bb481870039a

TODO

- [x] Experiment more with [available options / API](https://daybrush.com/moveable/release/latest/doc/Moveable.Snappable.html) for snappable
- [x] Fix drag group functionality (i.e. remove elements in current selection to avoid glitchy UX with snappable (remove on groupDragStart and re-add selected elements back on groupDragEnd))
- [x] Fix similar functionality for resize (we should remove current element from being considered in snappable)
- [x] Testing ... lots of testing
  - [x] Noticed that there are some scaling issues with pan / zoom (@drew08t any help with this would be 🙏)
- [x] Clean up
- [x] Docs
  - [x] Initial draft, not sure it makes sense to include unless we allow the user to turn this functionality off (i.e. just have what's new entry vs docs update) -> question: does it make sense to make this another canvas option to toggle on / off vs just defaulting to always being on?
- [x] What's new entry? (may be an easy win for 10.4, can be lumped with other canvas low hanging fruit updates)

### CODE REVIEW

File: docs/sources/panels-visualizations/visualizations/canvas/index.md [hunk 1]
```diff
@@ -107,6 +107,12 @@ When right clicking an element, you are able to edit, delete, duplicate, and mod
 
 {{< figure src="/static/img/docs/canvas-panel/canvas-context-menu-9-2-0.png" max-width="750px" caption="Canvas element context menu" >}}
 
+### Element snapping and alignment
+
+When moving elements around the canvas, snapping and alignment guides to help you create more precise layouts.
```
Comment (imatwawana) [inline docs/sources/panels-visualizations/visualizations/canvas/index.md:112]:
```suggestion
When you're moving elements around the canvas, snapping and alignment guides help you create more precise layouts.
```

**Review by imatwawana — COMMENTED** [review]
Small edit

File: public/app/features/canvas/runtime/scene.tsx [hunk 2]
```diff
@@ -426,9 +436,25 @@ export class Scene {
       .on('dragStart', (event) => {
         this.ignoreDataUpdate = true;
         this.setNonTargetPointerEvents(event.target, true);
+
+        if (this.moveable && this.moveable.elementGuidelines) {
+          const targetIndex = this.moveable.elementGuidelines.indexOf(event.target);
```
Comment (baldm0mma) [inline public/app/features/canvas/runtime/scene.tsx:441]: Nit: may not be worth it, but there's some duped code a few time here that we could maybe abstract out? Either way, this is awesome!

Comment (nmarrs) [reply public/app/features/canvas/runtime/scene.tsx:441]: ah yes that is a good point, a lot of repeated code here, thanks for pointing this out!

**Comment by drew08t** [comment]
Tested it locally and works great; however, as you mentioned, I am experiencing a couple issues during zoom, which I documented here: https://github.com/grafana/grafana/issues/81528

Because these appear to be caused by moveable itself, we will disable snappable during zoom for now until we can follow-up on this.

**Review by drew08t — APPROVED** [review]
Behaving nicely, with follow-up tasks documented with regards to zoom / multi-select.

File: public/app/features/canvas/runtime/scene.tsx (hunk 2 above)
Comment (nmarrs) [reply public/app/features/canvas/runtime/scene.tsx:442]: on closer review breaking out this logic I think is not really in scope of these changes (honestly the `scene.tsx` file has been needing a cleaning up for a while now (ideally breaking out moveable / selecto config to their own file where we can better group helper functions specific to setting up config (and also cleaning up `scene.tsx` a ton)

For now I've included some comments to help with future readability / maintainability 🤞

File: public/app/features/canvas/runtime/SceneTransformWrapper.tsx [hunk 3]
```diff
@@ -13,10 +13,37 @@ type SceneTransformWrapperProps = {
...
   const onZoom = (zoomPanPinchRef: ReactZoomPanPinchRef) => {
     const scale = zoomPanPinchRef.state.scale;
+    scene.scale = scale;
+  };
+
+  const onZoomStop = (zoomPanPinchRef: ReactZoomPanPinchRef) => {
+    const scale = zoomPanPinchRef.state.scale;
+    scene.scale = scale;
+    updateMoveable(scale);
+  };
```
Comment (nmarrs) [inline public/app/features/canvas/runtime/SceneTransformWrapper.tsx:23]: @drew08t i've done some minor cleanup to these changes, would appreciate a quick set of 👀 to make sure I'm not doing anything strange 😬

File: docs/sources/panels-visualizations/visualizations/canvas/index.md [hunk 4]
```diff
@@ -107,6 +107,12 @@ When right clicking an element, you are able to edit, delete, duplicate, and mod
 
 {{< figure src="/static/img/docs/canvas-panel/canvas-context-menu-9-2-0.png" max-width="750px" caption="Canvas element context menu" >}}
 
+### Element snapping and alignment
+
+When you're moving elements around the canvas, snapping and alignment guides help you create more precise layouts.
+
```
Comment (nmarrs) [inline docs/sources/panels-visualizations/visualizations/canvas/index.md:113]: should we add a note here that this feature only works when the panel is not zoomed in at all? (i.e. default scale) cc @drew08t

File: docs/sources/panels-visualizations/visualizations/canvas/index.md [hunk 5]
```diff
@@ -111,6 +111,10 @@ When right clicking an element, you are able to edit, delete, duplicate, and mod
 
 When you're moving elements around the canvas, snapping and alignment guides help you create more precise layouts.
 
+{{% admonition type="note" %}}
+For the time being, element snapping and alignment only works when the canvas is not zoomed in.
```
Comment (imatwawana) [inline docs/sources/panels-visualizations/visualizations/canvas/index.md:115]:
```suggestion
Currently, element snapping and alignment only works when the canvas is not zoomed in.
```

**Comment by Hipska** [comment]
Does this also include option to select multiple items and align them Left/Center/Right/Top/Middle/Bottom?

**Pull request merged**